### ISSUE TITLE
performance of QAT with openssl-3.0 / 3.1 vs. openssl-1.1

### ISSUE THREAD

**Issue by rdwr-vitalyk** [intel/QAT_Engine#236]
Has anyone compared performance of ssl offloading using  openssl-3.0 ?
we run test using  ECDHE-RSA-AES128-GCM-SHA256 and observe at least 30% degradation in transactions/second in comparison with same software based on openssl-1.1.

can you please advise where to look for the sourse of the problem?
thank you!

**Comment by Yogaraj-Alamenda** [comment]
Is the ECDH curve used is X25519 ? If yes, drop is expected as X25519 is not accelerated in OpenSSL3.0 engine due to known issue https://github.com/openssl/openssl/issues/19047

**Comment by vitalyk-radware** [comment]
i would like to ask about PRF offloading, please.
i see this code in e_qat.c:
# ifndef QAT_OPENSSL_3
     if (!ENGINE_set_pkey_meths(e, qat_pkey_methods)) {
          WARN("ENGINE_set_pkey_meths failed\n");
          goto end;
     }
# endif

i also see that in my test, functions such as qat_tls1_prf_ctrl are never called.
can you please tell why qat_pkey_methods are disabled in openssl v3 ?
Thank you!

**Comment by Yogaraj-Alamenda** [comment]
It is disabled due to known issues. We are working on fix in the qatengine and in the openSSL PR below to enable pkey methods in OpenSSL3.0 Engine. 

https://github.com/openssl/openssl/pull/20680

**Comment by vitalyk-radware** [comment]
when we compared performance of openssl 1.1 and 3.0 , we noticed  a significant degradation, and the most obvious possible explanation is that PRF is not done in hardware. Therefore we are looking for this feature. 
The  curve (X25519/X448) is not in our focus at the moment.
when you say "it might be possible PRF will still be not enabled", does it mean it is not going to be available is 3.0 / 3.1 at all?
can we do antything about it?  where we can learn  more about  these "known issues" with PRF offloading?
many thanks,
 Vitaly

**Comment by Yogaraj-Alamenda** [comment]
We are working on the fix/workarround at qatengine. I will keep you posted.

Below is the limitation at openSSL that I am talking about
https://github.com/openssl/openssl/issues/19047

**Comment by vitalyk-radware** [comment]
Thank you very much for your reply.
I read in this thread  that you said "There is provider and engine implementation in hardware. Our intention is to get OpenSSL default methods for fallback scenario". 
does it mean that we may get full QAT functionality (including PRF offload) if we load "qatengine" provider (instead of "default" or "legacy")  rather than load ENGINE "qat"?

thank you!

**Comment by Yogaraj-Alamenda** [comment]
@vitalyk-radware Yes Provider supports full functionality but we didnt test with any real time application apart OpenSSL's S_Server as we don't have any application (Nginx, Haproxy) that supports provider yet.

**Comment by gaozhangfei** [comment]
curious, does qat engine support both openssl 1.1 and openssl 3.0/3.1?
or openssl 1.1 only support qat engine while openssl 3.0 has to switch to qat provider?

**Comment by vitalyk-radware** [comment]
We use openssl 3 with Engine.
It is deprecated perhaps, but it works.

**Comment by gaozhangfei** [comment]
> We use openssl 3 with Engine. It is deprecated perhaps, but it works. 

Thanks @vitalyk-radware

**Comment by zhuguang11111** [comment]
Hi vitalyk,

May I ask have you fix the issue that PRF not offload for qatengine  api ?  
we are facing the same issue that when apply Openssl3 + QAT Engine: 
1) if choosing provider, but provider no support engine ctrl message , so no external polling etc.
2) if choosing engine api, the performance issue (prf not offload)

Thanks

**Comment by vitalyk-radware** [comment]
Hi ,
No , we have not solved these problems.
Regards,
Vitaly
